Support for Styled Components 3.0.0+

**General Information**

- [ ] Bug
- [ ] Improvement
- [ ] Feature
- [X] Other

**Description**

Add support for Styled Components V3 and up

Currently updating a projects' styled-components dependency above 3.0.0 results in the following error

```js
Error: Cannot find module 'styled-components/lib/models/StyleSheet'
    at Function.Module._resolveFilename (module.js:536:15)
    at Function.Module._load (module.js:466:25)
    at Module.require (module.js:579:17)
    at require (internal/module.js:11:18)
```

This is a result of SC only shipping flat bundles from V3, as outlined in their release notes: https://github.com/styled-components/styled-components/releases/tag/v3.0.1

Will submit a PR to add support shortly

**Steps to reproduce**

Update Styled Components to version 3.0.0 plus and run tests

**Versions**

- styled-components-test-utils: 0.3.1
- node: 8.10

- jest: N/A
- expect: N/A
- chai: 3.5.0
- jasmine: N/A

--- a/package.json
+++ b/package.json
@@ -91,7 +91,7 @@
     "react-test-renderer": "16.0.0",
     "rimraf": "2.6.1",
     "strip-ansi": "3.0.1",
-    "styled-components": "2.2.2",
+    "styled-components": "3.2.5",
     "webpack": "3.0.0"
   },
   "dependencies": {
@@ -106,7 +106,7 @@
     "strip-ansi": "^3.0.1",
     "react-test-renderer": "^16.0.0",
     "react": "^16.0.0",
-    "styled-components": "^2.2.2"
+    "styled-components": "^3.2.5"
   },
   "npmName": "styled-components-test-utils",
   "npmFileMap": [

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "styled-components-test-utils",
-  "version": "0.3.1",
+  "version": "0.4.0",
   "description": "Test utils for styled-components compatible with jest, expect, chai and jasmine",
   "main": "lib/index.js",
   "jsnext:main": "es/index.js",

--- a/src/init.js
+++ b/src/init.js
@@ -1,10 +1,10 @@
-import styleSheet from 'styled-components/lib/models/StyleSheet';
+import StyleSheet from './utils/styleSheet';
 import isOverV2 from './utils/isOverV2';
 import isServer from './utils/isServer';
 
 const init = () => {
   if (isOverV2()) {
-    styleSheet.reset(isServer());
+    StyleSheet.reset(isServer());
   }
 };
 

--- a/src/utils/getCSS.js
+++ b/src/utils/getCSS.js
@@ -1,6 +1,7 @@
 import { ServerStyleSheet } from 'styled-components';
-import StyleSheet from 'styled-components/lib/models/StyleSheet';
 import css from 'css';
+
+import StyleSheet from './styleSheet';
 import isOverV2 from './isOverV2';
 import isServer from './isServer';
 
@@ -11,7 +12,7 @@ const getStyle = (html) => {
   let matches;
 
   // eslint-disable-next-line
-  while((matches = STYLE_TAGS_REGEXP.exec(html)) !== null) {
+  while ((matches = STYLE_TAGS_REGEXP.exec(html)) !== null) {
     style += matches[1].trim();
   }
 

--- a/src/utils/styleSheet.js
+++ b/src/utils/styleSheet.js
@@ -15,7 +15,7 @@ const getStyleSheet = () => {
   }
 
   try {
-    // eslint-disable-next-line global-require
+    // eslint-disable-next-line global-require, import/no-unresolved
     return require('styled-components/lib/models/StyleSheet').default;
   } catch (e) {
     throw new Error('Could not find module styled-components/lib/models/StyleSheet.js');

--- a/src/utils/styleSheet.js
+++ b/src/utils/styleSheet.js
@@ -1,0 +1,24 @@
+import {
+  isStyledComponent,
+  __DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS,
+} from 'styled-components';
+
+const getStyleSheet = () => {
+  if (isStyledComponent) {
+    const secretInternals = __DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS;
+
+    if (secretInternals === undefined || secretInternals.StyleSheet === undefined) {
+      throw new Error('Could not find styled-components v3 secret internals');
+    }
+
+    return secretInternals.StyleSheet;
+  }
+
+  try {
+    return require.resolve('styled-components/lib/models/StyleSheet').default;
+  } catch (e) {
+    throw new Error('Could not find module styled-components/lib/models/StyleSheet.js');
+  }
+};
+
+export default getStyleSheet();

--- a/src/utils/styleSheet.js
+++ b/src/utils/styleSheet.js
@@ -15,7 +15,8 @@ const getStyleSheet = () => {
   }
 
   try {
-    return require.resolve('styled-components/lib/models/StyleSheet').default;
+    // eslint-disable-next-line global-require
+    return require('styled-components/lib/models/StyleSheet').default;
   } catch (e) {
     throw new Error('Could not find module styled-components/lib/models/StyleSheet.js');
   }

--- a/test/matchers/toMatchSnapshot.spec.js
+++ b/test/matchers/toMatchSnapshot.spec.js
@@ -12,9 +12,7 @@ type LinkProps = {
 };
 
 const toMatchSnapshot = (name, component) => {
-  expect(renderer.create(component).toJSON()).toMatchSnapshot(
-    `${name} - react-test-renderer`,
-  );
+  expect(renderer.create(component).toJSON()).toMatchSnapshot(`${name} - react-test-renderer`);
   expect(shallow(component)).toMatchSnapshot(`${name} - shallow`);
   expect(mount(component)).toMatchSnapshot(`${name} - mount`);
 };
@@ -54,10 +52,11 @@ test('basic', () => {
 });
 
 test('any component', () => {
-  const Link = ({ className, children }: LinkProps) =>
+  const Link = ({ className, children }: LinkProps) => (
     <a className={className}>
       {children}
-    </a>;
+    </a>
+  );
 
   const StyledLink = styled(Link)`
     color: palevioletred;
@@ -206,7 +205,5 @@ test('referring to other components', () => {
   expect(renderer.create(component).toJSON()).toMatchSnapshot(
     'referring to other components - react-test-renderer',
   );
-  expect(mount(component)).toMatchSnapshot(
-    'referring to other components - mount',
-  );
+  expect(mount(component)).toMatchSnapshot('referring to other components - mount');
 });

